### PR TITLE
Migrate at-apply utilites with template migrations

### PR DESCRIPTION
**Pull request by philipp-spiess**
This PR extracts all _candidate migrations_ from the existing _template migrations_ and reuses these in the `@apply` CSS migration. Seems like this _JustWorks✨_.

### CODE REVIEW

File: packages/@tailwindcss-upgrade/src/codemods/migrate-at-apply.test.ts [hunk 1]
```diff
@@ -68,3 +69,36 @@ it('should move the legacy `!` prefix, to the new `!` postfix notation', async (
     }"
   `)
 })
+
+it('should apply all candidate migration when migrating with a config', async () => {
```
Comment (adamwathan) [inline packages/@tailwindcss-upgrade/src/codemods/migrate-at-apply.test.ts:73]:
```suggestion
it('should apply all candidate migrations when migrating with a config', async () => {
```

File: packages/@tailwindcss-upgrade/src/codemods/migrate-at-apply.test.ts [hunk 2]
```diff
@@ -68,3 +69,36 @@ it('should move the legacy `!` prefix, to the new `!` postfix notation', async (
...
+        }),
+      )
+      .process(input, { from: expect.getState().testPath })
+      .then((result) => result.css)
+  }
+
+  expect(
+    await migrateWithConfig(css`
+      .foo {
+        @apply !tw_flex [color:--my-color];
```
Comment (adamwathan) [inline packages/@tailwindcss-upgrade/src/codemods/migrate-at-apply.test.ts:96]: Do we want to include an example of each template migration here?

File: packages/@tailwindcss-upgrade/src/codemods/migrate-at-apply.ts [hunk 3]
```diff
@@ -30,6 +36,12 @@ export function migrateAtApply(): Plugin {
       return [...variants, utility].join(':')
     })
 
+    // If we have a valid designSystem and config setup, we can run all
+    // candidate migrations on each utility
+    if (designSystem && userConfig) {
+      params = params.map((param) => migrateCandidate(designSystem, userConfig, param))
+    }
```
[the remark this answers is not in the document]
Comment (RobinMalfait) [reply packages/@tailwindcss-upgrade/src/codemods/migrate-at-apply.ts:43]: Yes, but I think a better way is to try to figure out where the config file lives (similar to how Tailwind CSS v3 just finds the config file if it's a known file name + extension in the root). If you have a custom file (aka, we can't resolve one) then the `--config` option should be required.

Another tricky thing, but will be more related to the stuff I'm working on with @thecrypticace (where we handle stylesheets as graphs to figure out which file imports what), is the fact that in theory you can have multiple roots with different settings. Which means that `@apply` in one file _could_ require `tw:` prefixes, and the other file might not require prefixes.

File: integrations/upgrade/index.test.ts [hunk 4]
```diff
@@ -83,7 +87,15 @@ test(
       `,
     )
 
-    await fs.expectFileToContain('src/input.css', css`@import 'tailwindcss' prefix(tw);`)
+    await fs.expectFileToContain('src/input.css', css` @import 'tailwindcss' prefix(tw); `)
+    await fs.expectFileToContain(
+      'src/input.css',
+      css`
+        .btn {
+          @apply tw:rounded-md! tw:px-2 tw:py-1 tw:bg-blue-500 tw:text-white;
```
Comment (RobinMalfait) [inline integrations/upgrade/index.test.ts:95]: It's so beautiful 🥹

**Pull request merged**